src/System.Private.CoreLib/shared/System/Number.Grisu3.cs

```diff
@@ -8,23 +8,34 @@ namespace System
 {
     internal static partial class Number
     {
+        // This is a port of the `Grisu3` implementation here: https://github.com/google/double-conversion/blob/a711666ddd063eb1e4b181a6cb981d39a1fc8bac/double-conversion/fast-dtoa.cc
+        // The backing algorithm and the proofs behind it are described in more detail here: http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf
+        // ======================================================================================================================================== 
+        //
+        // Overview:
+        //
+        // The general idea behind Grisu3 is to leverage additional bits and cached powers of ten to generate the correct digits.
+        // The algorithm is imprecise for some numbers. Fortunately, the algorithm itself can determine this scenario and gives us
+        // a result indicating success or failure. We must fallback to a different algorithm for the failing scenario.
         internal static unsafe class Grisu3
```

Comment:
What pieces are `unsafe` on this class? Would it make more sense to just mark the methods `unsafe` as opposed to the whole class?

Reply:
We have a bug tracking the general removal of `unsafe` in these methods and changing it to use `Span<T>` (https://github.com/dotnet/coreclr/issues/18162), we should do that cleanup there.

Reply:
Yes, but what about this class is still `unsafe` after your changes? It looks like you've changed all the pointers into Span.

---
In reply to: [252427656](https://github.com/dotnet/coreclr/pull/22040#discussion_r252427656) [](ancestors = 252427656)

Reply:
Ah, I will check.

Reply:
No more unsafe code exists. Removed.